src/core/server/saved_objects/migrationsv2/integration_tests/multiple_kibana_nodes.test.ts

```diff
@@ -0,0 +1,219 @@
+    migrations: {
+      '7.14.0': (doc) => {
+        if (doc.attributes?.status) {
+          doc.attributes.status = doc.attributes.status.replace('not_migrated', 'migrated');
+        }
+        return doc;
+      },
+    },
+  };
+
+  beforeAll(async () => {
+    await removeLogFiles();
+  });
+
+  afterAll(async () => {
+    await new Promise((resolve) => setTimeout(resolve, 10000));
+  });
+
+  beforeEach(async () => {
+    rootA = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_A.log`),
+    });
+    rootB = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_B.log`),
+    });
+    rootC = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_C.log`),
+    });
+
+    const { startES } = kbnTestServer.createTestServers({
+      adjustTimeout: (t: number) => jest.setTimeout(t),
+      settings: {
+        es: {
+          license: 'trial',
+          // original SO:
+          // [
+          //   { id: 'foo:1', type: 'foo', foo: { status: 'not_migrated_1' } },
+          //   { id: 'foo:2', type: 'foo', foo: { status: 'not_migrated_2' } },
+          //   { id: 'foo:3', type: 'foo', foo: { status: 'not_migrated_3' } },
+          // ];
+          dataArchive: Path.join(__dirname, 'archives', '7.13.0_5k_so.zip'),
+        },
+      },
+    });
+    esServer = await startES();
+  });
+
+  afterEach(async () => {
+    await Promise.all([rootA.shutdown(), rootB.shutdown(), rootC.shutdown()]);
+
+    if (esServer) {
+      await esServer.stop();
+    }
+  });
+
+  it('migrates saved objects normally when multiple Kibana instances are running', async () => {
+    const setupContracts = await Promise.all([rootA.setup(), rootB.setup(), rootC.setup()]);
+
+    setupContracts.forEach((setup) => setup.savedObjects.registerType(fooType));
+
+    const startContracts = await Promise.all([rootA.start(), rootB.start(), rootC.start()]);
+
+    const esClient = startContracts[0].elasticsearch.client.asInternalUser;
+    const migratedDocs = await fetchDocs(esClient, migratedIndex);
+
+    expect(migratedDocs.length).toBe(5000);
+    migratedDocs.forEach((doc, i) => {
+      expect(doc.id).toBe(`foo:${i}`);
+      expect(doc.foo.status).toBe(`migrated_${i}`);
+      expect(doc.migrationVersion.foo).toBe('7.14.0');
+    });
+  });
+
+  it('migrates saved objects normally when multiple Kibana instances are running with different plugins', async () => {
+    const setupContracts = await Promise.all([rootA.setup(), rootB.setup(), rootC.setup()]);
+
+    // Only register migration from one instance to simulate scenario
+    // where different plugins are enabled on different Kibana instances.
+    setupContracts[0].savedObjects.registerType(fooType);
+
+    // TODO: Fails if rootA.start promise isn't resolved first.
+    // Make sure this isn't an issue with the algorithm.
+    await rootA.start();
+    const startContracts = await Promise.all([rootB.start(), rootC.start()]);
```

Comment:
🚨  We need to sort this out before merging 🚨

This test is checking [test scenario 4](https://github.com/elastic/kibana/tree/master/src/core/server/saved_objects/migrationsv2#test-scenario-4-multiple-instances-mixed-plugin-enabled-configs) from the v2 migrations test plan.

`rootA` is the root instance that has the SO type registered to it.

Moving it inside the `Promise.all` with the others consistently returns the following error:

![Screen Shot 2021-05-14 at 2 55 27 PM](https://user-images.githubusercontent.com/1608770/118334284-7ab4e100-b4ca-11eb-99d2-ccff05984221.png)

This only happens if the following are true:
- `A` is started at the same time as `B` and `C` (inside `Promise.all`). Moving `A` outside and `await`ing it first (as I've done here to get tests to pass) resolves the issue.
- `A` has a different SO registered that `B` and `C` do not have. Registering the same SO to `B` and `C` resolves the issue.

I was unable to reproduce this issue manually after several attempts. Then I noticed that we don't consider this an officially supported upgrade scenario:

> Test scenario 4 (multiple instances, mixed plugin enabled configs):
>
> We don't support this upgrade scenario, but it's worth making sure we don't have data loss when there's a user error.

IIRC, don't we typically tell users to shut down all Kibana instances during upgrade, and then only restart them one at a time?

So I guess I have 3 questions:

1. Is this a supported upgrade scenario, where we start all Kibana instances at once?
2. If it is, am I doing something wrong in my test setup to get the above error? (I still consider this a likely possibility, despite spending quite awhile debugging) 😉 
3. If I'm not doing anything wrong, do we have a legit race condition in the algorithm?

Reply:
This is not an actual supported upgrade scenario in the sense that we don't guarantee that all Kibana nodes will function correctly when they are not configured in the same way.

The only think we should guarantee is that there isn't data loss and the `foo` documents are still present in the new index. I think we can safely ignore the exception and just guarantee that the docs are still present in the new index name.

Reply:
> Is this a supported upgrade scenario, where we start all Kibana instances at once?

Yes, according to our docs https://www.elastic.co/guide/en/kibana/current/upgrade-migrations.html#_upgrading_multiple_kibana_instances_2

> Then I noticed that we don't consider this an officially supported upgrade scenario:

According to the RFC, we do support this case until `8.0`
https://github.com/elastic/kibana/blob/master/rfcs/text/0013_saved_object_migrations.md#4214-handling-documents-that-belong-to-a-disabled-plugin

Reply:
> If I'm not doing anything wrong, do we have a legit race condition in the algorithm?

You can reproduce this problem if you comment out the 
```js
setupContracts[0].savedObjects.registerType(fooType);
```
Looking at the problem

Reply:
I think I only showed up on the git blame for this because I renamed the variable from `latestVersion` to `latestMigrationVersion` in my PR. The check existed before that 😄 

The assumption in the linked code:

```ts
// If we have a docVersion and the latestMigrationVersion is smaller than it or does not exist,
// we are dealing with a document that belongs to a future Kibana / plugin version.
```

In other words: if an object type is present and it is not registered with Kibana, then it will fail migration.

> Does it mean the migrations algorithm doesn't support the `orphan` SO object?

Correct.

I'm not as familiar with the migrationsv2 algorithm, but at the end of the day, if an unknown saved object type winds up being processed by the DocumentMigrator, it will fail. Perhaps we can update this to allow the document to pass through without being transformed iff the object type is not registered with Kibana. Thoughts?

Reply:
> but at the end of the day, if an unknown saved object type winds up being processed by the DocumentMigrator, it will fail.

Doesn't it contradict this comment at the top of the file https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/migrations/core/document_migrator.ts#L40 and the migration [Handling documents that belong to a disabled plugin](https://github.com/elastic/kibana/blob/master/rfcs/text/0013_saved_object_migrations.md#4214-handling-documents-that-belong-to-a-disabled-plugin) section of RFC?

Reply:
This error is happening because the saved object _does_ have a `migrationVersion` specified, but the object type is not registered with Kibana.

I don't think this contradicts the RFC: "Refuse to start a migration until the plugin is enabled or it's data deleted"
This appears to be behaving as intended. The migration fails because it detected an unknown object type, right?

Reply:
> "Refuse to start a migration until the plugin is enabled or it's data deleted"

The quote belongs to option (3). At the very bottom of this section:
```
We prefer option (3) since it provides flexibility for disabling plugins in the same version while also protecting users' data in all cases during an upgrade migration. However, because this is a breaking change we will implement (5) during 7.x and only implement (3) during 8.x.
```

Reply:
Ah I missed the last sentence :smile: I guess it does appear to contradict that. FWIW, I think migrationsv1 behaved the same way though. In migrationsv1, if an index migration is triggered, *all* documents in the index are fed to the DocumentMigrator and this error would happen, I believe.

Reply:
> In migrationsv1, if an index migration is triggered, all documents in the index are fed to the DocumentMigrator and this error would happen, I believe.

It seems so, but I'm not sure it's desired behavior. 
Unfortunately, our test coverage is not good enough to understand whether we support this case https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/migrations/core/document_migrator.test.ts#L307-L344

@pgayvallet any input on the matter?  

If we've never supported it and don't plan to do so, we should adjust RFC and the migration algorithm (do we need `OUTDATED_DOCUMENTS_*` steps then?).

Reply:
> If we've never supported it and don't plan to do so, we should adjust RFC and the migration algorithm (do we need `OUTDATED_DOCUMENTS_*` steps then?).

If we can verify that this behavior is/was happening in Migrations v1 for 7.11 (last version where v1 migrations was the default), then I agree we should just accept this behavior since we wanted it anyways in 8.0.

If it wasn't the behavior in 7.11 but we did ship this as the default behavior in 7.12 and 7.13...well I think we should probably accept it anyways. This ship has already sailed and I don't think trying to "fix" the behavior now and then change again in 8.0 is a good experience. We need migration behavior to be as stable as we feasibly can (which is why we need these tests :smile:). In 7.11 we also stopped shipping an OSS version of Kibana which helps mitigate against the most common case we know of for this scenario (the [Spaces example in the RFC](https://github.com/elastic/kibana/blob/master/rfcs/text/0013_saved_object_migrations.md#4214-handling-documents-that-belong-to-a-disabled-plugin)), and the config keys for disabling plugins are not widely documented.

If that is indeed the case, we'll need to make a few adjustments to v2 migrations. I believe we can drop #97965 and replace it with a task to remove the `OUTDATED_DOCUMENTS_*` steps @mshustov mentioned above. I'd be happy to take that on and defer on this PR until it's complete. I'll need to spend some more time thinking through other parts of the algo to be sure there aren't other adjustments that would need to be made as well.

Reply:
> If we can verify that this behavior is/was happening in Migrations v1 for 7.11 

It was. This issue is very old. I remember a discussion with Rudolf on a SDH one year ago where a customer disabled a plugin before performing a migration (security or space probably), and the migration was failing because of the 'unknown' SO types.

Reply:
Note that we *could* just remove the `!latestMigrationVersion` condition, and return `false` in that case (only throw if the version is effectively higher than the latest registered)

https://github.com/elastic/kibana/blob/26272faaf7bfa3d110cec7f1ccd1f694d7d56590/src/core/server/saved_objects/migrations/core/document_migrator.ts#L722-L728

I was a while ago since we discussed that with @rudolf, and I don't remember exactly why, but I think this had some subtle implications though.

Reply:
I'll do some testing on this tomorrow and verify if/how long it's been working this way.

Reply:
> I'll do some testing on this tomorrow and verify if/how long it's been working this way.

This may help: IIRC, the migrationv1 IndexMigrator specifically checks for objects that have an outdated `migrationVersion` field. So if your index only has an "orphaned" object, I don't think the IndexMigrator will trigger a full migration. However, if your index has an orphaned object _and_ a known object type that is outdated, then a full migration will be triggered for all documents in the index, and you'll see the error when the DocumentMigrator tries to process the orphaned object.

Reply:
After many permutations, I have been able to finally reproduce similar behavior across all versions since 6.8.16 (I did not try any earlier versions) up to the latest 7.14-SNAPSHOT using v2 migrations.

## Tl;dr

This behavior was essentially always present, but requires a few criteria to be triggered:
- An object with a `migrationVersion` must be present for a type that is not registered with Core any longer (due to a plugin being disabled)
- An index migration must be triggered.

Because of these two requirements, this behavior has gone likely unnoticed for quite some time. The most practical way to trigger this is:
- Start Kibana with all plugins enabled
- Create an object that has migrations added in a future version
- Upgrade Kibana to a later enough version to trigger index migrations, with spaces disabled

This situation has likely been increasing in frequency due to 2 factors:
- v1 Migrations: as more migrations have accumulated for more types as Kibana development has progressed, increasing the frequency of index migrations and possible orphan documents with a `migrationVersion` specified
- v2 Migrations always triggers an index migration on every version change

## Detailed notes of observed behavior
### **6.8.16 - 7.1.2**
Migrations will execute and fail with an error _also_ when all of the following conditions are met:
- An outdated document is detected (`migrationVersion` < latest registered migration for a given type)
- A document for unknown type exists
- *and* the rollup plugin is disabled

The reason this only happens when the rollup plugin is disabled is sort of by chance, but prior to 7.2 this plugin registered a migration for the `index-pattern` type for 6.5, making this SO type the only type during this version range where a migration the type registration and migration registration are not in sync. When the rollup plugin is disabled, the `index-pattern` migration is not registered, but there will be documents with the `migrationVersion` field specified if there were previous migrated.

If it weren't for this 'bug', this version range would behave the same as 7.2-7.11.2. This was fixed in 7.2: https://github.com/elastic/kibana/pull/31644

<details>
<summary>Reproduction steps</summary>

- Startup fresh Kibana 6.8.16 (default distribution), load sample ecommerce data
- Remove the `migrationVersion` field from a visualization with this command:
    ```sh
    curl -XPOST "http://localhost:9200/.kibana/_doc/visualization:37cc8650-b882-11e8-a6d9-e546fe2bba5f/_update" -H 'Content-Type: application/json' -d'
    {
      "doc": {
        "migrationVersion": null
      }
    }'
    ```
- Stop Kibana
- Startup Kibana 6.8.16 with `xpack.rollup.enabled: false`
- Kibana starts up and fails when attempting to apply an index migration:
  ```
  Error: Document "ff959d40-b880-11e8-a6d9-e546fe2bba5f" has property "index-pattern" which belongs to a more recent version of Kibana (6.5.0).
  ```

</details>

### **7.2.0 - 7.11.2**

Triggering this error in v1 migrations in this version range is a bit trickier because you must be sure to upgrade to late enough version to trigger an index migration, because in v1 migrations we only executed index migrations when needed. This can be 'forced' in a way however by modifying the `migrationVersion` field of a document.

Migrations will execute and fail with this error *only* when all of the following conditions are met:
- An outdated document is detected (`migrationVersion` < latest registered migration for a given type)
- A document for an unknown type exists that has a `migrationVersion` specified

<details>
<summary>Reproduction steps</summary>

- Startup fresh Kibana 7.2.0 (default distribution), load sample ecommerce data
  - This creates the default spaces document
- Set the `migrationVersion` field on a visualization to a previous migration version with this command to force an index migration:
    ```sh
    curl -XPOST "http://localhost:9200/.kibana/_doc/visualization:37cc8650-b882-11e8-a6d9-e546fe2bba5f/_update" -H 'Content-Type: application/json' -d'
    {
      "doc": {
        "migrationVersion": {
          "visualization": "7.0.0"
        }
      }
    }'
    ```
- Stop Kibana
- Startup Kibana 7.2.0 with `xpack.spaces.enabled: false`
- Migrations fail when attempting to apply an index migration:
  ```
  Error: Document "default" has property "space" which belongs to a more recent version of Kibana (6.6.0).
  ```

</details>


### 7.12 - master (v2 migrations)

Migrations will execute and fail with this error *only* when all of the following conditions are met:
- A document for an unknown type exists that has a `migrationVersion` specified
- An index migration has been triggered, which happens on every version change

This scenario will not happen if you simply disable a plugin that registers an SO type that has existing documents in the index. You must actually do a version upgrade as well to trigger this scenario because it only happens during index migrations.

<details>
<summary>Reproduction step</summary>

- Startup fresh Kibana 7.13.0 (default distribution)
  - This creates the default spaces document
- Stop Kibana
- Startup Kibana 7.14.0-SNAPSHOT with `xpack.spaces.enabled: false`
- Migrations fail when attempting to apply an index migration:
  ```
  Error: Document "default" has property "space" which belongs to a more recent version of Kibana (6.6.0).
  ```

</details>

## Conclusions

The reason this test fails is because it triggers an index migration due to the version diff between the index in the archive (7.12.0) and the version that Kibana is running with (8.0.0). For the nodes without the type registered, this causes a failure.

I think since this has been the behavior in a growing number of situations, we should probably just commit to it at this point and move forward with implementing option (3) from [section 4.2.1.4 of the RFC](https://github.com/elastic/kibana/blob/master/rfcs/text/0013_saved_object_migrations.md#4214-handling-documents-that-belong-to-a-disabled-plugin). If so, I think we can get rid of the outdated document search as well, as mentioned above.

Thoughts?

Reply:
Providing `This behavior was essentially always present`, the most practical solution might be to commit to (3) explicitly. It requires logging the name of plugin-owner of these `orphan` documents if possible and improving the errors message to make it actionable (`remove SO id: XXX or enable plugin Foo back`).

Reply:
> we should probably just commit to it at this point and move forward with implementing option (3) from [section 4.2.1.4 of the RFC](https://github.com/elastic/kibana/blob/master/rfcs/text/0013_saved_object_migrations.md#4214-handling-documents-that-belong-to-a-disabled-plugin)

This makes sense to me.

So back to this test then: considering one of the scenarios we wanted to cover was `Add a test to start nodes with different plugins enabled`, does it even make to move forward with this test at this point, as we will ultimately be expecting migrations to not even start in this scenario?

We probably have 2 options for next steps:
1. merge with the first test that's been written and exclude the second one altogether
2. update the second test to expect a failure and wait to merge this PR until we've implemented the desired changes in the algorithm

WDYT?

Reply:
> merge with the first test that's been written and exclude the second one altogether

I'd go with this option. 
The second case might be tested in the unit tests. IMO It's not important enough to spend 30sec. to spin up the test env.

Reply:
+1 let's move forward with this PR without the second test. I've opened a new issue to track the changes we need to do to commit to option (3), though I have some concerns about that option as well. We can discuss it here: https://github.com/elastic/kibana/issues/101052

Reply:
Done, thanks all for the input. This PR should be ready for a final review.